kolibri/core/tasks/validation.py

```diff
@@ -19,6 +23,31 @@ def validate(self, data):
             "extra_metadata": {},
         }
 
+    def validate_for_restart(self, job):
+        """
+        :param job: The job for which to restart
+        :type job: kolibri.core.tasks.job.Job
+        :return: A dictionary of data for instantiating a new job
+        """
+        if job.state not in [State.CANCELED, State.FAILED]:
+            raise serializers.ValidationError(
+                "Cannot restart job with state={}".format(job.state)
+            )
+
+        return {
+            # default behavior is to retain the same job ID, so the existing job requires deletion
+            "job_id": job.job_id,
+            "args": copy.copy(job.args),
+            "kwargs": copy.copy(job.kwargs),
+            "track_progress": job.track_progress,
+            "cancellable": job.cancellable,
+            "extra_metadata": job.extra_metadata.copy(),
+            "facility_id": job.facility_id,
+        }
+
+    def to_representation(self, instance):
```

Comment:
It took me a little while to work through in my head why this was doing something in the case of a restart, but not in any other instance.

I wonder if it might be better to not override `to_representation` here, and just explicitly call `validate_for_restart` in the API endpoint restart method to get the final data that we want?

Reply:
The hacking of a REST serializer for job validation is what makes this awkward, in my opinion.

Reply:
It's for validating data coming from the API endpoint, so I think it's a reasonable use case - any suggestions for how to update the way we're using it to make it less awkward? I was not a biggest fan of having to return and update the dict of the job args, but I couldn't think of another way while implementing it.

Reply:
It's fairly uncoupled from the API endpoints, so its implementation doesn't seem to enforce use solely through API requests. If that's the intent, it seems that the logic in `Task.validate_job_data` would be better placed within the `TasksViewSet.validate_create_req_data` method.

Otherwise, if we keep `validate_job_data`, then I think removing inheritance of `serializers.Serializer` and defining only the functionality we intend, purely validation related logic, will clarify the implementation. The `to_representation` method is an unimplemented method in `BaseSerializer` but has an implementation in the `Serializer` which seems we don't intend to utilize, per your comments. That would remove the validator's use case of accepting an instance, as opposed to a data object, which is the functionality that I hacked given the current structure.

Reply:
> It's fairly uncoupled from the API endpoints, so its implementation doesn't seem to enforce use solely through API requests.

This is also a feature of DRF Serializers in general - they can be used for deserialization/validation outside of an API endpoint.

> If that's the intent, it seems that the logic in Task.validate_job_data would be better placed within the TasksViewSet.validate_create_req_data method.

There is one place it is used outside of an API endpoint - in migrating the `validate_sync_task` function that it replaced, it is used not to process data from an API endpoint within the single user syncing functions.

> I think removing inheritance of serializers.Serializer and defining only the functionality we intend, purely validation related logic, will clarify the implementation

This seems fine - I had not done this previously to avoid vendoring/duplicating code from DRF, but I think as long as it is using the serializer MetaClass there shouldn't be any gotchas here.

I can file a follow up issue for cleanup to avoid this being blocking here!

Reply:
> they can be used for deserialization/validation outside of an API endpoint.

And serialization, like I've done here :)

Reply:
> And serialization, like I've done here :)

Hah - true - maybe just some code comments to sign post it would be enough, I just had to think longer than I thought I should have to realize.